libyul/backends/evm/EVMDialect.cpp

```diff
@@ -450,35 +476,49 @@ SideEffects EVMDialect::sideEffectsOfInstruction(evmasm::Instruction _instructio
 	};
 }
 
-BuiltinFunctionForEVM const* EVMDialect::verbatimFunction(size_t _arguments, size_t _returnVariables) const
+BuiltinHandle EVMDialect::verbatimFunction(size_t _arguments, size_t _returnVariables) const
 {
-	std::pair<size_t, size_t> key{_arguments, _returnVariables};
-	std::shared_ptr<BuiltinFunctionForEVM const>& function = m_verbatimFunctions[key];
-	if (!function)
-	{
-		BuiltinFunctionForEVM builtinFunction = createFunction(
-			"verbatim_" + std::to_string(_arguments) + "i_" + std::to_string(_returnVariables) + "o",
-			1 + _arguments,
-			_returnVariables,
-			SideEffects::worst(),
-			std::vector<std::optional<LiteralKind>>{LiteralKind::String} + std::vector<std::optional<LiteralKind>>(_arguments),
-			[=](
-				FunctionCall const& _call,
-				AbstractAssembly& _assembly,
-				BuiltinContext&
-			) {
-				yulAssert(_call.arguments.size() == (1 + _arguments), "");
-				Expression const& bytecode = _call.arguments.front();
-
-				_assembly.appendVerbatim(
-					asBytes(formatLiteral(std::get<Literal>(bytecode))),
-					_arguments,
-					_returnVariables
-				);
-			}
-		).second;
-		builtinFunction.isMSize = true;
-		function = std::make_shared<BuiltinFunctionForEVM const>(std::move(builtinFunction));
-	}
-	return function.get();
+	auto const it = std::find_if(m_verbatimFunctions.begin(), m_verbatimFunctions.end(), [&](BuiltinFunctionForEVM const& _function) {
+		return _function.numParameters == 1 + _arguments && _function.numReturns == _returnVariables;
+	});
+	yulAssert(_arguments <= verbatimMaxInputSlots);
+	yulAssert(_returnVariables <= verbatimMaxOutputSlots);
+	if (it != m_verbatimFunctions.end())
+		return {static_cast<size_t>(std::distance(m_verbatimFunctions.begin(), it))};
+
+	BuiltinFunctionForEVM builtinFunction = createFunction(
+		"verbatim_" + std::to_string(_arguments) + "i_" + std::to_string(_returnVariables) + "o",
+		1 + _arguments,
+		_returnVariables,
+		SideEffects::worst(),
+		std::vector<std::optional<LiteralKind>>{LiteralKind::String} + std::vector<std::optional<LiteralKind>>(_arguments),
+		[=](
+			FunctionCall const& _call,
+			AbstractAssembly& _assembly,
+			BuiltinContext&
+		) {
+			yulAssert(_call.arguments.size() == (1 + _arguments), "");
+			Expression const& bytecode = _call.arguments.front();
+
+			_assembly.appendVerbatim(
+				asBytes(formatLiteral(std::get<Literal>(bytecode))),
+				_arguments,
+				_returnVariables
+			);
+		}
+	);
+	builtinFunction.isMSize = true;
+	m_verbatimFunctions.push_back(std::move(builtinFunction));
+	yulAssert(m_verbatimFunctions.size() <= verbatimIdOffset);
+	return {m_verbatimFunctions.size() - 1};
+}
+
+BuiltinFunctionForEVM const& EVMDialect::builtinFunction(BuiltinHandle const& handle) const
+{
+	if (handle.id < verbatimIdOffset)
+		return m_verbatimFunctions.at(handle.id);
```

Comment:
I mean, you already have `Handles` which looks kinda like an attempt to make some lookups faster? Not sure why it contains only a few builtins. Also, it only works for the static lookups of known builtins, while dynamic ones are a concern too (especially in the parser and analysis).

Reply:
> Won't that result in some non-determinism? Their order could differ based on which sources were compiled before.

That would be the case for any numerical id scheme that doesn't order by hash of the string identifier as it is right now but assigns based on occurrence. If you think this is a problem then we should have a discussion about the numerical ids approach in general, I think. We could of course take care that, e.g., block hasher and such don't use the ids but fall back onto the strings as well as prohibit that the ids are used in sets and maps - but that would kind of defeat the purpose.

> This means that creation is no longer transparent. 

True, good catch! I'll get rid of the mutable.

> I think that we might be better off storing the verbatim functions in a map, indexed with either their arities or their names.

My assumption is that there won't be more than in order of 10 verbatim functions defined, so having a map lookup would be slow and wasteful vs. using a dense vector that only contains the ones that were used. Or so was my reasoning when going for this approach :)

> I mean, you already have Handles which looks kinda like an attempt to make some lookups faster? Not sure why it contains only a few builtins.

Contains the ones that are needed throughout the codebase, I didn't want to add superfluous ones.

>  Also, it only works for the static lookups of known builtins, while dynamic ones are a concern too (especially in the parser and analysis).

It was clear to me that the linear lookup would be slower than what we had previously (the slowdown you mentioned below), so I bundled these in the dialect so the handles are cached and can be used directly without having to have another lookup. This of course doesn't work for dynamic ones in the same way. The helper map should help, though. :)

Reply:
> That would be the case for any numerical id scheme that doesn't order by hash of the string identifier as it is right now but assigns based on occurrence.

In a general case, but here we have a limited space of 10000 possible IDs. You could deterministically assign those IDs with a rule like `ins * 100 + outs`. Then pretend they always exist and generated them on access by handle.

You'd just need to use a `map` rather than a `vector` to store them.

> having a map lookup would be slow and wasteful vs. using a dense vector that only contains the ones that were used. 

I'm not sure if this difference will matter in practice. I mean, you're right that it's wasteful to make it logarithmic complexity when it can be constant, but if the downside is introducing non-determinism, then I think that logarithmic is more than acceptable (it could also be amortized O(1) with `unordered_map` though). I suspect that it will matter way less than the logarithmic vs linear we were discussing in the other comment.

I'd suggest to check benchmarks. If the difference is measurable, we can still use a `vector` and just preallocate all items - with just 10k of them it's still viable.

> Contains the ones that are needed throughout the codebase, I didn't want to add superfluous ones.

What confused me is that not all of the omitted ones are actually superfluous. Some of them are used too. This should have a comment explaining which ones are supposed to be added there.

Reply:
> What confused me is that not all of the omitted ones are actually superfluous. Some of them are used too. This should have a comment explaining which ones are supposed to be added there.

Right, I guess it's not immediately clear. Also `Handles` isn't super descriptive, although I can't really think of a better name right now, either. What I meant to put there was handles which are not accessed through the inherited accessors (`discardFunction`, `equalityFunction` etc) but used in a non-dynamic fashion (e.g. when constructing these ghost calls with `eq`) to avoid repeated lookups by string identifier.